extensions/amp-a4a/0.1/template-renderer.js

```diff
@@ -22,27 +22,29 @@ import {getAmpAdTemplateHelper} from './template-validator';
  */
 export class TemplateRenderer extends FriendlyFrameRenderer {
```

Comment:
in general, inheritance like this is discouraged. Let's extract shared logic to a util method instead, and have `TemplateRenderer extends Renderer `. See [here](https://en.wikipedia.org/wiki/Composition_over_inheritance) for more background around why.

Reply:
Instead of extending FriendlyFrameRenderer, can I keep an instance of it in TemplateRenderer? Since TemplateRenderer eventually renders into a friendly frame, it's unavoidable that I'll need the entirety of FriendlyFrameRenderer's logic.

Reply:
totally. i suggest convert FriendlyFrameRenderer into a util class without extending Renderer.

Reply:
For now I'd prefer to keep it as its own Renderer subclass, so its consistent and uniform with the other renderers. This will especially be useful once I finally get around to tackling the amp-a4a refactor.

Reply:
how about a structure like:

TemplateRenderer -> Renderer
AmpRenderer (naming TBD) -> Renderer

Both renderer import the friendlyframeutil to share code.

Reply:
My concern is that the util would just be the renderer itself, as there's little to nothing that would be left after extracting the common code. Can we circle back to this in a future PR?

Reply:
Having one renderer impl extend another renderer impl is the code pattern i want to avoid. Let's avoid at the beginning.

Given it's introduced in a previous PR, I'm OK to have this PR in first. But please do refactor in the next PR.

Reply:
Sure, I completely agree with you on avoiding a `TemplateRenderer -> FriendlyFrameRenderer -> Renderer` hierarchy -- that's why I changed TemplateRenderer to use an instance of FriendlyFrameRenderer rather than extend it.

But, yes, we can do a refactor/cleanup after this PR. Meanwhile, feel free to start reviewing this.